src/sql/src/plan/statement/ddl/connection.rs

```diff
@@ -164,17 +168,61 @@ impl ConnectionOptionExtracted {
+                        "ACCESS KEY ID option is required for AWS Connection with credentials"
+                    ),
+                    (Some(_), None, _) => sql_bail!(
+                        "SECRET ACCESS KEY option is required for AWS Connection with credentials"
+                    ),
+                    (None, None, Some(_)) => sql_bail!(
+                        "ACCESS KEY ID and SECRET ACCESS KEY is required for AWS Connection with credentials"
+                    ),
+                    _ => None,
+                };
+
+                let assume_role = match (self.assume_role_arn, self.assume_role_session_name) {
+                    (Some(arn), session_name) => {
+                        let mz_arn =
+                            scx.catalog.aws_external_connection_role().ok_or_else(|| {
+                                sql_err!(
+                                    "Could not find AWS external connection role in the catalog"
+                                )
+                            })?;
+                        let external_id_prefix =
+                            scx.catalog.aws_external_id_prefix().ok_or_else(|| {
+                                sql_err!("Could not find AWS External ID prefix in the catalog")
+                            })?;
+                        Some(AwsAuth::AssumeRole(AwsAssumeRole {
+                            arn,
+                            session_name,
+                            mz_principal: MaterializePrincipal {
+                                arn: mz_arn,
+                                external_id_prefix,
+                            },
+                        }))
+                    }
+                    (None, Some(_)) => {
+                        sql_bail!("ASSUME ROLE ARN is required for AssumeRole AWS Connection")
+                    }
+                    _ => None,
+                };
+
+                if credentials.is_some() && assume_role.is_some() {
+                    sql_bail!("Either ACCESS KEY ID and SECRET ACCESS KEY credentials or ASSUME ROLE ARN should be provided for AWS Connection, not both");
+                }
+
                 Connection::Aws(AwsConfig {
-                    credentials: AwsCredentials {
-                        access_key_id: self
-                            .access_key_id
-                            .ok_or_else(|| sql_err!("ACCESS KEY ID option is required"))?,
-                        secret_access_key: self
-                            .secret_access_key
-                            .ok_or_else(|| sql_err!("SECRET ACCESS KEY option is required"))?
-                            .into(),
-                        session_token: self.token,
-                    },
+                    auth: credentials.or(assume_role).ok_or_else(|| {
+                        sql_err!("Provide either ACCESS KEY ID and SECRET ACCESS KEY credentials or ASSUME ROLE ARN for AWS Connection")
+                    })?,
```

Comment:
We tend to prefer structured errors to `sql_bail!` and `sql_err!`.

Reply:
Done!

Reply:
I will do a follow up PR to make use of the new errors at other places if possible.